ctr/src/ctr32.rs

```diff
@@ -1,249 +0,0 @@
-//! Generic implementation of CTR mode with a 32-bit counter
-//! (big or little endian), generic over block ciphers.
-
-use cipher::{
-    block::{Block, BlockCipher, BlockEncrypt, ParBlocks},
-    generic_array::{typenum::Unsigned, ArrayLength, GenericArray},
-    stream::{FromBlockCipher, LoopError, SyncStreamCipher},
-};
-use core::{convert::TryInto, marker::PhantomData, mem};
-/// CTR mode with a 32-bit big endian counter.
-///
-/// Used by e.g. AES-GCM.
-pub struct Ctr32BE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<GenericArray<u8, B::BlockSize>>,
-    Block<B>: Copy,
-{
-    ctr: Ctr32<B, BigEndian>,
-}
-
-/// CTR mode with a 32-bit little endian counter.
-///
-/// Used by e.g. AES-GCM-SIV.
-pub struct Ctr32LE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<GenericArray<u8, B::BlockSize>>,
-    Block<B>: Copy,
-{
-    ctr: Ctr32<B, LittleEndian>,
-}
-
-impl<B> FromBlockCipher for Ctr32BE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<Block<B>>,
-    Block<B>: Copy,
-{
-    type BlockCipher = B;
-    type NonceSize = B::BlockSize;
-
-    #[inline]
-    fn from_block_cipher(cipher: B, nonce: &Block<B>) -> Self {
-        Self {
-            ctr: Ctr32::new(cipher, *nonce),
-        }
-    }
-}
-
-impl<B> FromBlockCipher for Ctr32LE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<Block<B>>,
-    Block<B>: Copy,
-{
-    type BlockCipher = B;
-    type NonceSize = B::BlockSize;
-
-    #[inline]
-    fn from_block_cipher(cipher: B, nonce: &Block<B>) -> Self {
-        let mut counter_block = *nonce;
-        counter_block[15] |= 0x80;
```

Comment:
Note this is a tricky implementation detail of how AES-GCM-SIV defines a 32-bit counter mode. Perhaps it should be hoisted back into the `aes-gcm-siv` crate.

Reply:
IIUC it will be enough to simply modify nonce in the `aes-gcm-siv` crate before passing it to the CTR type, no?

Reply:
Yep. Perhaps I can try to move it into the `aes-gcm-siv` crate before we cut another release of `ctr`

Reply:
An alternative solution would be to define a GCM-SIV flavor with `load` method modifying the nonce as specified. This flavor can be either defined in the `ctr` crate, or in `aes-gcm-siv`. But I think modifying nonce before passing to the constructor will be a simpler solution.

Reply:
Removing the high bit masking in this PR: https://github.com/RustCrypto/stream-ciphers/pull/197